### PR TITLE
CheMaster3000 rewrite with TGUI

### PR DESCRIPTION
**Pull request by Saicchi**
[CHEMISTRY][QOL][TGUI][ENHANCEMENT][SILICONS][CLEAN]

## About the PR 

Reworks the CheMaster300 giving it a (hopefully) nice UI, cleaner code and some new features.
- Choose bottle type:
	- Merges vials, small bottles, bottles and all can types into this option.
		- Default selection is the 50u biodegradeable bottle.
		- All bottle types can not be recycled.
- Choose multiple pill types:
	- Default selection is the vanilla pill which takes whichever color the chemical mixture is.
	- Variety of designs of floor pills for the clown!
- **Cyborgs can now use the ChemMaster3000**
- Quality of Life:
	- Put items and beakers in your hand or designated spot on the floor.
	- **Beaker automatically ejects when empty**.
	- Drag and Drop valid containers in the machine.
	- Quick switch beakers when inserting another one.
- **Minimum chemical amount for anything is 1 unit instead of 5 units (everything is rounded so no fractions).**
	- Might be a point of concern.
		- Item spam.
		- 100 1u pills from a beaker instead of 20u pills.

I'd appreciate if people tested the PR and gave me some feedback on stuff and things I missed.

This PR works only on ChemMaster 3000, meaning the other two devices ( mini chemaster3000 and prototype chemaster ) will not be affected.

image
image
image
image

The analyze button now shows the message on the chat.
image

## Why's this needed? 

ChemMaster 3000 is the only of the main 4 chemistry machines that doesn't have TGUI.

## Changelog

```changelog
(u)DrWolfy
(*)CheMaster3000 rework with TGUI.
```

### CODE REVIEW

**Comment by DisturbHerb** [comment]
pretty sure that the biodegradable bottle was a balance change, currently (and rightly, i guess) you can only make glass bottles/vials in the recycler

**Comment by Saicchi** [comment]
> EDIT - i'm not quite sure why the chemaster should be able to make a non-recyclable version of the glass bottles/vials

Vials have a unique injection method so I can understand that. The 30u glass bottle could easily be replaced with a 30u plastic bottle, provided someone makes a sprite for it.

I only added both to maintain the functionality.

File: tgui/packages/tgui/interfaces/ChemMaster.js [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+            title="Select bottle icon">
+            {
+              bottle_icons.map(button_image)
+            }
+          </SectionEx>)
+      }
+      {
+        icon_list === ICON_LIST_PATCHES && (
+          <SectionEx capitalize
+            title="Select patch icon">
```
Comment (Azrun) [inline tgui/packages/tgui/interfaces/ChemMaster.js:168]: Use something other than the term "icon", like style or type. That was it maintains emersion preferably.

**Review by Azrun — COMMENTED** [review]
Really wish the new features weren't part of this PR and the features were atomized for review a and comment.

**Review by ZeWaka — CHANGES_REQUESTED** [review]
'operations' window is way too large compared to the reagents preview area

you don't need all that explainer text in the operations window, we have tooltips and button styling for a reason

**Comment by Saicchi** [comment]
image
image
image

How about this? Increased the reagent list size, removed most of the tooltips and reduced width and height.

**Comment by ZeWaka** [comment]
> How about this? Increased the reagent list size, removed most of the tooltips and reduced width and height. Also removed the icon terminology and went with type instead.

1. I'd make the container icons clearly look like buttons. Right now they look just like images with a tooltip.
2. probably less space between each row of options, it's still feeling super crammed - maybe make the reagent display smaller?

**Comment by Saicchi** [comment]
> 1. I'd make the container icons clearly look like buttons. Right now they look just like images with a tooltip.
They had the background button on my first design I posted on discord, I made it transparent due to user feedback.

> 2. probably less space between each row of options, it's still feeling super crammed - maybe make the reagent display smaller?
Reverted the reagent display to the normal height ( used everywhere that ReagentList is used ).

I don't know how I could compress things any more since each row in the operations is already operating with no margin or padding.
image

The ChemMaster screen is already quite small, here is a full game image with the Reagent Extractor for comparision.

image

**Comment by Saicchi** [comment]
> Technically, there's this margin(?) here

All rows are the same width and column since they are using `<Stack>`. Removing it would make them uneven and rather ugly IMO.
I guess it could be slimmed down if the `Use bottle` and `Use patch box` text were moved to the side of the button instead of bottom.

>  it'd be best if you update the main pr body with the new images

Will do.

**Comment by Saicchi** [comment]
Margin can go negative! This time it really can't reduce the height much further without affecting the selecting bottle window. 

image
image

**Comment by pali6** [comment]
I don't see any DM-side TGUI code here, am I missing something?

**Comment by Saicchi** [comment]
> I don't see any DM-side TGUI code here, am I missing something?

Github is having a hard time showing a good diff with its git diff algorithm.

https://github.com/goonstation/goonstation/blob/44703b28bb3583cd2cade90d29385b9da45bbc96/code/modules/chemistry/Chemistry-Machinery.dm#L457-L814

**Comment by vortex1942** [comment]
image
Teeny tiny thing, but the padding between the flush/eject button is a lot smaller when compared to the other buttons on the menu.

**Comment by ZeWaka** [comment]
btw the weird icon offsets have been fixed in master

**Comment by Saicchi** [comment]
Indeed. They look much neater now.
image

**Comment by TobleroneSwordfish** [comment]
Sorry this has been inactive for so long, but since I've recently gained the ability to hit the funny merge button I thought I'd look over it.
A few minor issues still:
- Clicking on the machine with your hand doesn't open the interface, you have to actually swap the beaker to access it.
- The pill and bottle icons are rendered at half in-game size making the pills almost impossible to see, have you tried seeing what they look like at normal in-game resolutions? (use the pixelated flag on `<Icon>` to prevent blurring)
- The reset button looks like it should reset the entire operations section and is also unnecessary since the input box has a reset button built in.
- The `Remove Amount` buttons seem largely pointless and are completely different to how any other chemistry machine works, consider just having `isolate` and `remove all` for parity with the reagent extractor.

**Comment by Cyle99** [comment]
Thanks for the feedback. I'll take a look into it.

**Comment by TobleroneSwordfish** [comment]
- The icon issue isn't a big deal if it would be a pain to rework the UI to render them at full size.
- The reset button can just be removed imo since the textbox has one built in.
- There isn't a lot of reason for the chemmaster to need to be that precise at removing chems but if you want it to do that then at least copy/componentize the chem dispenser controls so you don't have two different looking UIs that do the same thing.

**Comment by pali6** [comment]
I'm not much of a fan of lowering the minimum amount to 1u. That will lead to even worse spam of pills etc.

**Comment by Saicchi** [comment]
I've set the minimum amount for anything at 5 units.
Removed the reset button.

image

> There isn't a lot of reason for the chemmaster to need to be that precise at removing chems but if you want it to do that then at least copy/componentize the chem dispenser controls so you don't have two different looking UIs that do the same thing.

They're already more of less the same. The ChemMaster is already using the same component for reagents from the Reagent Extractor.

ChemDispenser
image

ChemMaster
image
Reagent Extractor
image

**Comment by TobleroneSwordfish** [comment]
!merge_upstream

**Pull request merged**